### PR TITLE
Add returnMergedEntity option

### PR DESCRIPTION
**Pull request by mshima**
Please make sure the below checklist is followed for Pull Requests.
  - [ ] Checks are green
  - [ ] Tests are added where necessary
  - [ ] Documentation is added/updated where necessary
  - [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/jhipster-core/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: lib/exporters/jhipster_deployment_exporter.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ function checkForErrors(deployment) {
 }
 
 function setUpDeploymentStructure(deployment) {
-  const cleanDeployment = JSON.parse(JSON.stringify(deployment, null, 2));
+  const cleanDeployment = JSON.parse(JSON.stringify(deployment, null, 2).concat('\n'));
```
Comment (MathieuAA) [inline lib/exporters/jhipster_deployment_exporter.js]: The `cleanDeployment` name wasn't great... let's just remove the variable which is only used once and do l.54:
`deploymentToExport[GENERATOR_NAME] = JSON.parse(JSON.stringify(deployment, null, 2).concat('\n'));`

File: lib/exporters/jhipster_application_exporter.js [hunk 2]
```diff
@@ -49,10 +49,12 @@ function exportApplications(applications) {
  * @param application, the JDL application to exporters.
  * @return the exported application in its final form.
  */
-function exportApplication(application) {
+function exportApplication(application, configuration = {}) {
```
Comment (MathieuAA) [inline lib/exporters/jhipster_application_exporter.js]: I like it, passing it in a configuration object is a good idea!

File: lib/exporters/jhipster_entity_exporter.js [hunk 3]
```diff
@@ -92,7 +93,9 @@ function exportEntities(passedConfiguration) {
   if (shouldFilterOutEntitiesBasedOnMicroservice()) {
     filterOutEntitiesByMicroservice();
   }
-  writeEntities(passedConfiguration.application.forSeveralApplications ? configuration.application.name : '');
+  if (!passedConfiguration.configuration || !passedConfiguration.configuration.skipEntityFilesGeneration) {
```
Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js]: Might as well create a variable for that `!passedConfiguration.configuration || !passedConfiguration.configuration.skipEntityFilesGeneration`, `skippedEntityFileGenerations`

Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js]: Can you update the doc for this attribute `skipEntityFilesGeneration` please?

File: lib/exporters/jhipster_application_exporter.js [hunk 4]
```diff
@@ -49,10 +49,12 @@ function exportApplications(applications) {
...
   checkForErrors(application);
   const exportableApplication = setUpApplicationStructure(application);
-  writeConfigFile(exportableApplication);
+  if (!configuration.skipYoRcGeneration) {
```
Comment (MathieuAA) [inline lib/exporters/jhipster_application_exporter.js]: can you add the `configuration` & `skipYoRcGeneration` attributes to the doc please?

Comment (mshima) [reply lib/exporters/jhipster_application_exporter.js]: I was testing back there, forgot to add when I create the PR.
Added to JSDoc, but used a new format, can use a old format if you want.

Comment (MathieuAA) [reply lib/exporters/jhipster_application_exporter.js]: Nop that's good!

File: lib/exporters/jhipster_application_exporter.js [hunk 5]
```diff
@@ -46,13 +46,17 @@ function exportApplications(applications) {
 
 /**
  * Exports JDL a application to a JDL file in the current directory.
- * @param application, the JDL application to exporters.
+ * @param {Object}  application, the JDL application to exporters.
```
Comment (MathieuAA) [inline lib/exporters/jhipster_application_exporter.js]: typo: to export

File: lib/exporters/jhipster_application_exporter.js [hunk 6]
```diff
@@ -46,13 +46,17 @@ function exportApplications(applications) {
...
+ * @param {Boolean} [configuration.skipYoRcGeneration], set true to skip writing .yo-rc.json.
  * @return the exported application in its final form.
  */
-function exportApplication(application) {
+function exportApplication(application, configuration = {}) {
   checkForErrors(application);
   const exportableApplication = setUpApplicationStructure(application);
-  writeConfigFile(exportableApplication);
+  if (!configuration.skipYoRcGeneration) {
+    writeConfigFile(exportableApplication);
```
Comment (MathieuAA) [inline lib/exporters/jhipster_application_exporter.js]: this needs testing too

File: lib/exporters/export_utils.js [hunk 7]
```diff
@@ -36,14 +36,19 @@ function writeConfigFile(config, yoRcPath = '.yo-rc.json') {
...
         ...yoRc[GENERATOR_NAME],
-        ...config[GENERATOR_NAME]
+        ...config[GENERATOR_NAME],
+        creationTimestamp
       }
     };
   }
-  fs.writeFileSync(yoRcPath, JSON.stringify(newYoRc, null, 2));
+  fs.writeFileSync(yoRcPath, JSON.stringify(newYoRc, null, 2).concat('\n'));
 }
```
Comment (MathieuAA) [inline lib/exporters/export_utils.js]: this needs testing too

File: lib/jdl/jdl_importer.js [hunk 8]
```diff
@@ -184,7 +184,7 @@ function importOneApplicationAndEntities(jdlObject, configuration) {
     exportedDeployments: []
   };
   const jdlApplication = jdlObject.applications[Object.keys(jdlObject.applications)[0]];
-  importState.exportedApplications.push(JHipsterApplicationExporter.exportApplication(jdlApplication));
+  importState.exportedApplications.push(JHipsterApplicationExporter.exportApplication(jdlApplication, configuration));
```
Comment (MathieuAA) [inline lib/jdl/jdl_importer.js]: this needs testing too

Comment (mshima) [reply lib/jdl/jdl_importer.js]: I couldn't think test for importOneApplicationAndEntities, exportApplication is duplicated

Comment (MathieuAA) [reply lib/jdl/jdl_importer.js]: oh okay! I'll see about that, thanks

File: lib/exporters/jhipster_application_exporter.js (hunk 6 above)
Comment (mshima) [reply lib/exporters/jhipster_application_exporter.js]: Done

File: lib/exporters/export_utils.js (hunk 7 above)
Comment (mshima) [reply lib/exporters/export_utils.js]: Done

File: test/spec/exporters/jhipster_entity_exporter.spec.js [hunk 9]
```diff
@@ -767,7 +767,7 @@ describe('JHipsterEntityExporter', () => {
           fs.rmdirSync('.jhipster');
         });
 
-        it('creates the entity in the current folder', () => {
+        it("don't creates the entity in the current folder", () => {
```
Comment (MathieuAA) [inline test/spec/exporters/jhipster_entity_exporter.spec.js]: doesn't create

File: test/spec/exporters/jhipster_application_exporter.spec.js [hunk 10]
```diff
@@ -181,7 +205,7 @@ describe('JHipsterApplicationExporter', () => {
           fs.unlinkSync(path.join('.yo-rc.json'));
         });
 
-        it('adds the read content to the exported application', () => {
+        it('adds the read content to the exported application and keep selected', () => {
```
Comment (MathieuAA) [inline test/spec/exporters/jhipster_application_exporter.spec.js]: 2 things:
  - what does keep selected mean?
  - if there's an "and" in an "it", it usually means you're testing two things and this should be split in two "it"

File: test/spec/exporters/jhipster_application_exporter.spec.js [hunk 11]
```diff
@@ -148,6 +148,28 @@ describe('JHipsterApplicationExporter', () => {
...
+              enableTranslation: false,
+              languages: ['en', 'fr'],
+              jhipsterVersion: '4.9.0',
+              otherModules: ['MyModule']
+            }),
+            { skipYoRcGeneration: true }
+          );
+        });
+
+        it("don't creates the .yo-rc.json in the current folder", () => {
```
Comment (MathieuAA) [inline test/spec/exporters/jhipster_application_exporter.spec.js]: doesn't create*

File: test/spec/exporters/jhipster_application_exporter.spec.js [hunk 12]
```diff
@@ -175,13 +199,7 @@ describe('JHipsterApplicationExporter', () => {
...
             })
           );
-          content = JSON.parse(fs.readFileSync(path.join('.yo-rc.json'), { encoding: 'utf8' }));
-        });
-        after(() => {
-          fs.unlinkSync(path.join('.yo-rc.json'));
-        });
-
-        it('adds the read content to the exported application', () => {
+          const content = JSON.parse(fs.readFileSync(path.join('.yo-rc.json'), { encoding: 'utf8' }));
```
Comment (MathieuAA) [inline test/spec/exporters/jhipster_application_exporter.spec.js]: I don't like having logic inside an it, "it"s should only have "expect"s, and setup code inside a before/beforeEach

File: test/spec/exporters/jhipster_application_exporter.spec.js [hunk 13]
```diff
@@ -221,6 +239,35 @@ describe('JHipsterApplicationExporter', () => {
...
+              applicationType: MONOLITH,
+              baseName: 'toto',
+              packageName: 'com.mathieu.sample',
+              enableTranslation: false,
+              languages: ['en', 'fr'],
+              jhipsterVersion: '4.9.0',
+              creationTimestamp: 'new'
+            })
+          );
+          const content = JSON.parse(fs.readFileSync(path.join('.yo-rc.json'), { encoding: 'utf8' }));
```
Comment (MathieuAA) [inline test/spec/exporters/jhipster_application_exporter.spec.js]: this should be inside a before

**Comment by MathieuAA** [comment]
Can you finish this PR please @mshima?

**Comment by mshima** [comment]
@MathieuAA sure. I will try later today.

**Comment by mshima** [comment]
Probably I will split in smaller PRs.

**Comment by MathieuAA** [comment]
Thanks :) Do as you please!

**Comment by mshima** [comment]
@MathieuAA first round is ready.

**Comment by MathieuAA** [comment]
Can we close this one then?

**Comment by mshima** [comment]
> Can we close this one then?

There are 2 more changes.

**Comment by mshima** [comment]
@MathieuAA this is the last one for now.

File: .gitignore [hunk 14]
```diff
@@ -7,9 +7,9 @@ out/
...
 .sonar
-coverage/*
-.nyc_output/*
+coverage/
+.nyc_output/
 yarn-error.log
 jhipster-core.iml
 .project
-.vscode/*
+.vscode/
```
Comment (MathieuAA) [inline .gitignore]: Not part of the PR, you can directly merge to master for this, it's easier

File: lib/exporters/jhipster_entity_exporter.js [hunk 15]
```diff
@@ -124,7 +123,13 @@ function writeEntities(subFolder) {
   for (let i = 0, entityNames = Object.keys(configuration.entities); i < entityNames.length; i++) {
     const filePath = path.join(subFolder, toFilePath(entityNames[i]));
     const entity = updateEntityToGenerateWithExistingOne(filePath, configuration.entities[entityNames[i]]);
-    fs.writeFileSync(filePath, JSON.stringify(entity, null, 2).concat('\n'));
+    if (configuration.returnMergedEntity) {
```
Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js]: Shouldn't it always be true? I mean, I don't see cases where we don't want that...

File: lib/exporters/jhipster_entity_exporter.js [hunk 16]
```diff
@@ -124,7 +123,13 @@ function writeEntities(subFolder) {
...
+      // Set the updated entity back.
+      configuration.entities[entityNames[i]] = entity;
+    }
+    if (!configuration.skipEntityFilesGeneration) {
```
Comment (MathieuAA) [inline lib/exporters/jhipster_entity_exporter.js]: Having a conditional that says "if we don't want to write entity files" in a function that's called `writeEntities` is weird, don't you think?
Having two functions makes more sense.

File: lib/exporters/jhipster_entity_exporter.js (hunk 15 above)
Comment (mshima) [reply lib/exporters/jhipster_entity_exporter.js]: Your call.
I did for backward compatibility.
But since this can be considered a bug, a minor release should be enough for this change.
The only use case I can think is if some blueprint wants to change the behavior. But I don’t think we should support that.

File: lib/exporters/jhipster_entity_exporter.js (hunk 16 above)
Comment (mshima) [reply lib/exporters/jhipster_entity_exporter.js]: I agree but if we follow the rules, this change should be a major release.

File: lib/exporters/jhipster_entity_exporter.js (hunk 15 above)
Comment (MathieuAA) [reply lib/exporters/jhipster_entity_exporter.js]: Ah! good point. You're absolutely right! Let's make it the default, we'll do release including this later (I still have to finish the "options in app" thing)

File: lib/exporters/jhipster_entity_exporter.js (hunk 16 above)
Comment (MathieuAA) [reply lib/exporters/jhipster_entity_exporter.js]: No need, it's a private function. Had it been in the exports then yes.
How do you see it? Maybe we're not thinking about the same things :)

My idea is : `exportEntities` should call two functions. The first one being the "we want to write files to the disk" and the other "the flag is set, let's not write"

Comment (mshima) [reply lib/exporters/jhipster_entity_exporter.js]: I am thinking about the following approaches:
- Create a new mergeExistingEntityFromFile.
Duplicate current writeEntities removing the write part.
- Rename the function to mergeAndWriteEntities and don’t duplicate the loop

**Comment by mshima** [comment]
I just noticed that the only method that take subFolder (application.name) into account is writeEntities.
Am I wrong or should we change that?

**Comment by MathieuAA** [comment]
> I just noticed that the only method that take subFolder (application.name) into account is writeEntities.
Am I wrong or should we change that?

@mshima sorry didn't see your comment... Anyway, this is good to go, thanks :)
